File: c++/1463.cpp
#include <iostream>
#include <algorithm>

using namespace std;
int DP[1000001];
int N;
int main(){
    ios::sync_with_stdio(0);
    cin.tie(0);
    DP[1] = 0;
    DP[2] = 1;
    cin >> N;
    for(int i=3 ; i<=N ; i++){
        DP[i] = DP[i-1] + 1;
        if(i%2 == 0){
            DP[i] = min(DP[i],DP[i/2]+1);
        }
        if(i%3 == 0){
            DP[i] = min(DP[i],DP[i/3]+1);
        }
    }
    cout << DP[N];
    return 0;
}

File: python/1463.py
N = int(input())

DP = [0 for i in range(1000001)]

DP[2] = 1
DP[3] = 1
DP[4] = 2
DP[5] = 3

for i in range(6, 1000001):
    DP[i] = DP[i-1] + 1
    if i % 2 == 0:
        DP[i] = min(DP[i], DP[i//2] + 1)
    if i % 3 == 0:
        DP[i] = min(DP[i], DP[i//3] + 1)

print(DP[N])
